### ISSUE TITLE
PR #10 follow-up: address Apple MPS review findings

### ISSUE THREAD

**Issue by mandrean** [mandrean/ktransformers#11]
## Context

Follow-up umbrella issue for [PR #10](https://github.com/mandrean/ktransformers/pull/10) and the Apple `local_chat` investigation in [issue #9](https://github.com/mandrean/ktransformers/issues/9).

This issue tracks the review findings that currently keep PR #10 from being technically honest and approval-ready.

Full native MPS `balance_serve` remains out of scope for this remediation branch and stays tracked by the existing Apple/native backlog issues.

### Apple `local_chat` correctness
- [P1] The Apple TinyLlama `local_chat` / GGUF path at PR #10 head (`3faca35`) still produces non-finite prefill/decode logits and `<unk>`-style output.
  - Likely root cause at PR head: rotary runtime buffers such as `inv_freq` remain on `meta` after GGUF/meta injection.
  - Relevant files: `kt-sft/ktransformers/optimize/optimize.py`, `kt-sft/ktransformers/models/modeling_llama.py`, `kt-sft/tests/test_apple_deterministic_smoke.py`, `kt-sft/scripts/compare_tinyllama_logits.py`, `kt-sft/scripts/trace_tinyllama_prefill.py`
- [P2] Apple docs currently overclaim the state of support on the PR head by presenting TinyLlama `local_chat` as working before the integrated path is actually numerically healthy.
  - Relevant file: `doc/en/mps.md`

### Apple backend/operator correctness
- [P1] `KExpertsApple` uses `unsorted_outputs.index_copy_()` on MPS, which is not implemented on the stock backend targeted by this PR and causes the Apple acceleration suite to fail.
  - Relevant file: `kt-sft/ktransformers/operators/experts.py`
- [P2] `test_mps_optimize_rules_use_apple_backends()` incorrectly assumes every MPS YAML must contain `KLinearApple`, even though `TinyLlama-smoke-torch.yaml` is an intentional Torch-only bisection rule.
  - Relevant file: `kt-sft/tests/test_apple_acceleration.py`

### CI / workflow truthfulness
- [P2] `apple-native-acceleration.yml` only validates microtests and does not cover the integrated Apple TinyLlama `local_chat` smoke/debug path.
  - Relevant file: `.github/workflows/apple-native-acceleration.yml`
- [P1] `balance-serve-apple.yml` currently gives false confidence by validating the Python `sched_ext.py` shim via `PYTHONPATH`, not an actually built native extension.
  - Relevant file: `.github/workflows/balance-serve-apple.yml`

### Native `balance_serve` scope / gating
- [P2] `setup.py` detects native `balance_serve` dependencies from the wrong root (`kt-sft/third_party` instead of the repo-root `third_party` tree used by the restored CMake files), which can silently force fallback to the Python shim.
  - Relevant file: `kt-sft/setup.py`
- [P1] The restored native `balance_serve` CMake tree is not macOS-portable as checked in. It hardcodes `g++-11/12/13`, overwrites `CMAKE_CXX_COMPILER`, and forces non-portable flags.
  - Relevant file: `kt-sft/csrc/balance_serve/CMakeLists.txt`
- [P1] The restored native `sched_ext` / `kvc2` tree still lacks a real MPS backend. It remains CUDA/NPU-oriented and should not be represented as validated Apple-native support.
  - Relevant files: `kt-sft/csrc/balance_serve/sched/*`, `kt-sft/csrc/balance_serve/kvc2/*`

### Apple `local_chat` correctness
- [ ] Land the RoPE/runtime-buffer materialization fix after GGUF/meta injection
- [ ] Keep deterministic Apple smoke mode greedy and warning-free
- [ ] Ensure one-shot/piped `local_chat` exits cleanly without `EOFError`
- [ ] Keep TinyLlama compare/trace/debug tests green and finite
- [ ] Align `doc/en/mps.md` with the validated Apple smoke path

### Apple backend/operator correctness
- [ ] Replace the unsupported MPS `index_copy_` path in `KExpertsApple`
- [ ] Narrow `test_mps_optimize_rules_use_apple_backends()` so it excludes intentional diagnostic Torch-only rules
- [ ] Keep the Apple acceleration suite green on stock MPS without relying on global CPU fallback

### CI / workflow truthfulness
- [ ] Extend Apple CI to include the integrated TinyLlama smoke/debug path
- [ ] Rename or rescope the Apple `balance_serve` workflow so it only claims what it actually validates
- [ ] Avoid green workflows that imply native Apple support when only Python fallback is being exercised

### Native `balance_serve` scope correction
- [ ] Fix native dependency detection to use the real repo-root `third_party` layout
- [ ] Gate native `balance_serve` claims/builds to actually supported backends
- [ ] Treat Apple as Python fallback-only unless/until a real native MPS backend exists
- [ ] Keep the restored archive tree out of the "working Apple-native backend" story unless it builds and runs

### Docs alignment
- [ ] Keep `doc/en/mps.md` focused on the validated TinyLlama Apple smoke path
- [ ] Document the current native `balance_serve` story honestly for Apple
- [ ] Close this issue with a summary of what was fixed, what was downgraded, and what remains backlog work

**Comment by mandrean** [comment]
Progress update on `mandrean/apple-pr10-review-fixes`:

Fixed on this branch:
- Apple TinyLlama `local_chat` correctness is restored.
  - RoPE runtime buffers are now materialized after GGUF/meta injection.
  - deterministic smoke mode no longer relies on the old `top_k=0` workaround.
  - one-shot/piped `local_chat` exits cleanly instead of throwing `EOFError`.
  - TinyLlama smoke/debug coverage now passes with finite logits and sane text output.
- Apple acceleration test coverage is green again.
  - `KExpertsApple` no longer uses the unsupported MPS `index_copy_` path.
  - the optimize-rule test now treats `TinyLlama-smoke-torch.yaml` as an intentional diagnostic rule rather than a production Apple backend rule.
- CI/workflow scope is now more honest.
  - `apple-native-acceleration.yml` now downloads TinyLlama smoke-test assets and runs the integrated Apple TinyLlama smoke/debug suites on macOS instead of only microtests.
  - `balance-serve-apple.yml` has been renamed/scoped as a Python fallback workflow and explicitly asserts that it is validating `sched_ext.py`, not a native extension.
- Apple `balance_serve` install/gating is now explicit.
  - `setup.py` now resolves `third_party` from the repo root.
  - Apple/MPS is treated as Python-fallback-only for `balance_serve`; `KTRANSFORMERS_FORCE_BALANCE_NATIVE=1` now errors instead of implying a native Apple backend exists.
- Apple docs were updated to say that `balance_serve` on Apple currently uses the Python fallback, not a validated native MPS backend.

Validated locally on this branch:
- `KTRANSFORMERS_TINYLLAMA_MODEL_PATH=... KTRANSFORMERS_TINYLLAMA_GGUF_PATH=... PYTHONPATH=kt-sft python -m pytest -q kt-sft/tests/test_tinyllama_prefill_debug.py kt-sft/tests/test_apple_deterministic_smoke.py`
  - `16 passed`
- `PYTHONPATH=kt-sft python -m pytest -q kt-sft/tests/test_fp8_safetensor_dequant.py kt-sft/tests/test_apple_acceleration.py kt-sft/tests/test_balance_serve_sched_ext.py kt-sft/tests/test_issue2_runtime_paths.py`
  - `22 passed, 1 skipped`

Still intentionally not claimed as fixed here:
- the restored native `csrc/balance_serve` CMake tree is still not a validated macOS-portable build
- the restored native `sched_ext` / `kvc2` tree still does not provide a real MPS backend

Those remain backlog/native-scope items rather than something this remediation branch pretends to ship.

**Comment by mandrean** [comment]
The remaining native Apple `balance_serve` / `sched_ext` work is moving to [issue #7](https://github.com/mandrean/ktransformers/issues/7), which now has the full implementation plan.

This keeps #11 focused on the PR #10 remediation/status work while the true native-MPS scheduler/cache backend follow-up happens on the dedicated native Apple backlog thread.
